src/librustc_lint/unused.rs

```diff
@@ -270,47 +270,52 @@ impl UnusedParens {
                                 msg: &str,
                                 struct_lit_needs_parens: bool) {
         if let ast::ExprKind::Paren(ref inner) = value.node {
-            let necessary = struct_lit_needs_parens &&
-                            parser::contains_exterior_struct_lit(&inner);
-            if !necessary {
-                let span_msg = format!("unnecessary parentheses around {}", msg);
-                let mut err = cx.struct_span_lint(UNUSED_PARENS,
-                                                  value.span,
-                                                  &span_msg);
-                // Remove exactly one pair of parentheses (rather than naïvely
-                // stripping all paren characters)
-                let mut ate_left_paren = false;
-                let mut ate_right_paren = false;
-                let parens_removed = pprust::expr_to_string(value)
-                    .trim_matches(|c| {
-                        match c {
-                            '(' => {
-                                if ate_left_paren {
-                                    false
-                                } else {
-                                    ate_left_paren = true;
-                                    true
-                                }
-                            },
-                            ')' => {
-                                if ate_right_paren {
-                                    false
-                                } else {
-                                    ate_right_paren = true;
-                                    true
-                                }
-                            },
-                            _ => false,
-                        }
-                    }).to_owned();
-                err.span_suggestion_short_with_applicability(
-                    value.span,
-                    "remove these parentheses",
-                    parens_removed,
-                    Applicability::MachineApplicable
-                );
-                err.emit();
+            if struct_lit_needs_parens &&
+                    (parser::contains_exterior_struct_lit(&inner) ||
+                        if let ast::ExprKind::Break(..) = inner.node {
```

Comment:
You can probably simplify this addition down to `... || inner.node == ast::ExprKind::Break`

Reply:
That would be surprising to say the least, because `ExprKind::Break` is defined as a tuple struct of two values (note the two dots). I didn't know you could compare with incomplete value definitions.

Reply:
Sorry yes you may still need the two dots! What I was trying to point out is that you can just check if `inner.node` is a `ExprKind::Break` as the _whole_ condition and get ride of the `... { true } else { false }`.

Reply:
How would that work? The `if let` needs some values in the then/else branches.

Reply:
Ah yes I see what you mean. I was under the impression we could get a `bool` from `inner.node == ast::ExprKind::Break(..)`. What I was mainly hoping to simplify in all this was getting rid of the additional `... { true } else { false }` as it just looks a little odd to read.

A way I see doing that is keeping things how they were (with `necessary`) and going for something like adding those additional four lines to check for a `Break`.
```
let necessary = struct_lit_needs_parens && parser::contains_exterior_struct_lit(&inner);
if !necessary {
+    // We don't want to lint when there is a `break` in `inner`
+    if let ast::ExprKind::Break(..) = inner.node {
+        return
+    }
    ...
}
```
Either way - just an initial reaction I had to seeing the new additions! I'm fine keeping it how it is.